voting_ml/feature_selection.py

```diff
@@ -128,38 +202,119 @@ def ftsel_chi2(self, data, KBest = 'all', label=None, input_test_size=0, input_r
+        X_train = data_dict["X_train"].astype(str)
+        X_test = data_dict["X_test"].astype(str)
+        y_train = data_dict["y_train"].astype(str)
+        y_test = data_dict["y_test"].astype(str)
+
+        # prepare input data
+        print("X_train",X_train)
+        X_train_enc, X_test_enc, oe = self.prepare_inputs(X_train, X_test)
+
+        # prepare output data
+        y_train_enc, y_test_enc, le = self.prepare_targets(y_train, y_test)
+    
+        pca = PCA()
+        pca.fit(X_train_enc)
+        transform_matrix = pca.components_.T
+        eigen_vals = pca.explained_variance_
+
+        #Creating directory for the output
+        if(not os.path.isdir("../output/"+self._run_name)):
+            os.mkdir("../output/"+self._run_name)
+        
+        print("Eigen values for first 20 principal components: ",eigen_vals[0:20])
+        rank_eigenval = np.arange(1,X_train_enc.shape[1]+1).reshape((X_train_enc.shape[1]))
+        plt.plot(rank_eigenval,eigen_vals, label = "eigen-values")
+        plt.xlabel('Rank of the eigenvalue')
+        plt.ylabel('Eigen Value')
+        plt.yscale('log')
+        plt.title("PCA - eigenvalues")
+        plt.legend()
+        image_name = '../output/'+self._run_name+'/eigenvalues_vs_rank.png'
+        plt.savefig(image_name)
+        plt.clf()
+
+        total_var = np.zeros((eigen_vals.shape[0]))
+        total_sum = np.sum(eigen_vals)
+        for i in range(eigen_vals.shape[0]):
+            total_var[i] = np.sum(eigen_vals[0:i+1])*100/total_sum
+
+        num_95perVar = None
+        num_99perVar = None
 
-    def ft_corr(self, data, questions):
+        for j in range(eigen_vals.shape[0]):
+            if(total_var[j]>=95):
+                num_95perVar = j+1
+                break
+
+        for k in range(eigen_vals.shape[0]):
+            if(total_var[k]>=99):
+                num_99perVar = k+1
+                break
+
+        print("Number of principal components needed to represent 95% of the total variance: ", num_95perVar)
+        print("Number of principal components needed to represent 99% of the total variance: ", num_99perVar)
+
+        pca_data = data
+        pca_X_train = np.dot(X_train_enc,transform_matrix)
+        pca_X_test = np.dot(X_test_enc,transform_matrix)
+
+        pca_data_dict = {
+            'data': pca_data,
+            'X_train': pca_X_train[:,0:KBest],
+            'X_test': pca_X_test[:,0:KBest],
+            'y_train': y_train,
+            'y_test': y_test
+        }
+        
+        return pca_data_dict, transform_matrix, eigen_vals
+    
+    def ft_corr(self, X_train, questions):
+        if(questions==None):
+            fts = X_train.shape[1]
+            questions_int = list(map(str, list(range(1,fts+1,1))))
+            questions = ["ft_"+x for x in questions_int]
         KBest = len(questions)
-        X = data[:, :-1]
-        df = pd.DataFrame(X, columns = questions)
+        df = pd.DataFrame(X_train, columns = questions)
```

Comment:
What's the point of introducing a pandas dataframe here? You iterate over the columns later, but you could easily do that with array slicing and integers. It's a bit awkward to use a dataframe when you don't need to.

Reply:
the corr function is for dataframe. This is what I had found. I can look for a correlation function for numpy array. But the reason was I found this function.

Reply:
Oh, okay. I actually missed the call to `df.corr` in my pass earlier. We can probably just leave this as is then.